### PR TITLE
feat: add support for `oneOf` in sttp4 client

### LINKED ISSUE #19891: [BUG][Scala] OneOf generates incorrect classes.
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

When generating a model using `scala-sttp`, `oneOf` in the OpenAPI generates case classes that can never match any actual valid json.
I also tried some of the other scala generators and they seem to have the same problem.

##### openapi-generator version

Latest master

##### OpenAPI declaration file content or url

[Link to OpenAPI yaml gist](https://gist.github.com/GerretS/f2c8f98f05c5ca921144f56585192842)

##### Generation Details

```sh
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i /tmp/test/example.yaml -g scala-sttp -o /tmp/test/out
```

##### Steps to reproduce

If you look in the model package you will see that the specific animal types (`Dog`, `Cat`) are all perfectly fine.
However, the top-level Example case class has a field of type `Animal`:

```scala
case class Example(
  requestId: Option[String] = None,
  animal: Option[Animal] = None
)
```

And here is the generated Animal case class:
```scala
case class Animal(
  dogName: String,
  foodType: Option[CapybaraFoodType] = None,
  catId: Int,
  capyBaraId: Option[Int] = None
)
```

There are two problems with this:
- `dogName` and `catId` are only required in `Dog` and `Cat` respectively. This means that it is impossible to deserialize an `Example` containing a `Dog` unless you provide a `catId` for the `Dog`.
- the `foodType` is set to `CapybaraFoodType` even though `Dog` and `Cat` each have their own FoodType schema. An `Example` containing a `Cat` or `Dog` with a `CatFoodType` or `DogFoodType` cannot be modelled at all, even though it's valid according to the spec.

##### Suggest a fix

The most straightforward fix would be to make `Animal` an empty trait, and have `Cat`, `Dog` and `Capybara` be subtypes of this trait. The `Example` then contains the `Animal`.  Then, any (de)serializer could use the `discriminator` in the yaml to make an instance of the correct subtype.

Note that a trait is better than an abstract class in this case, because the OpenAPI spec allows an object to be reused among multiple oneOf statements for different objects, and a case class can extend multiple traits but only one abstract class.

I realize there might be more work to get the compatible serializer libs to work correctly with this new change, but since this case currently isn't supported at all. I would be very happy with just being able to generate a valid model as a first step.

### PR DESCRIPTION
**Pull request by plaflamme**
Address #19891 for sttp4 clients.

@fish86

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [WSL](https://learn.microsoft.com/en-us/windows/wsl/install))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [ ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR solves a reported issue, reference it using [GitHub's linking syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) (e.g., having `"fixes #123"` present in the PR description)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

---
## Summary by cubic
Adds oneOf support to the Scala `scala-sttp4` client with discriminator mapping, wrapper composition for shared members, and aligned JSON/request/response handling. Updates docs, adds a `scala-sttp4-circe` sample and CI, bumps dependencies, fixes parent-with-properties and aliased-child cases, and adds tests (including keyword discriminator wire-name). Fixes #19891.

- New Features
  - oneOf: sealed-trait parents; inline or wrapper composition for children; discriminator + mapping; uses discriminator.propertyBaseName for correct wire field (handles Scala keywords); discriminator fields removed in children; empty-member handling; aliased children resolved; parent props enforced on the trait and propagated to each child; custom circe/json4s codecs; total constructor-name mapping; tests for edge cases.
  - API alignment: body(asJson(...)); Either[ResponseException[String], T]; pass Option[String] directly to .header; query/form containers encode via mkString(","); url-form-encoded bodies include only provided fields; default Seq params = Seq.empty; AnyType/object map to circe Json/JsonObject or json4s JValue/JObject.
  - Enums/samples/deps: enums use sealed-trait ADTs with per-type codecs (json4s JsonSupport now uses enum companion Serializer objects); new `scala-sttp4-circe` sample + CI; sttp client 4.0.15; Scala 2.13.18; add `circe-generic` and `circe-generic-extras`; use semiauto derivation.

- Migration
  - Enums are sealed traits; update matches/usages and imports.
  - oneOf parents are sealed traits; parent properties appear on the parent trait and each child; wrappers may exist under the parent companion; adjust usages/imports if referenced directly.
  - Response type is Either[ResponseException[String], T] (was ResponseException[String, Exception]).
  - Query/form containers now use mkString(","); url-form-encoded bodies omit unset fields; verify server expectations.

<sup>Written for commit 21efd98e5b4cebbf0af0aa10cdaafcdb81a52b04. Summary will update on new commits.</sup>

### CODE REVIEW

**Comment by wing328** [comment]
thanks for the PR

cc @clasnake (2017/07), @shijinkui (2018/01), @ramzimaalej (2018/03), @chameleon82 (2020/03), @Bouillie (2020/04) @fish86 (2023/06)

**Comment by wing328** [comment]
https://github.com/OpenAPITools/openapi-generator/actions/runs/21774387501/job/62832048821?pr=22916

please follow step 3 to update the samples

File: modules/openapi-generator/src/main/resources/scala-sttp4/model.mustache [hunk 1]
```diff
@@ -15,46 +20,224 @@ import {{import}}
...
+{{#vendorExtensions.x-use-discr}}
+  // oneOf with discriminator - using semiauto derivation with Configuration
+  import io.circe.{Encoder, Decoder}
+  import io.circe.generic.extras._
+  import io.circe.generic.extras.semiauto._
+  import io.circe.generic.auto._
+  
+  private implicit val config: Configuration = Configuration.default.withDiscriminator("{{discriminator.propertyName}}")
+    .copy(
+      transformConstructorNames = {
```
[the remark this answers is not in the document]
Comment (plaflamme) [reply modules/openapi-generator/src/main/resources/scala-sttp4/model.mustache:194]: fixed in https://github.com/OpenAPITools/openapi-generator/pull/22916/commits/d05761d0b1f19f1a7ca092c2e67fa7d09045b6fe

**Comment by plaflamme** [comment]
@wing328 I've updated the samples in https://github.com/OpenAPITools/openapi-generator/pull/22916/commits/e59402398ece4b7baff3e08637360f4c65129acf

**Comment by wing328** [comment]
https://github.com/OpenAPITools/openapi-generator/actions/runs/21850815146/job/63223036192?pr=22916

please replace tabs with 4-space in the java file.

**Comment by plaflamme** [comment]
@wing328 should be all good now

File: bin/configs/scala-sttp4-circe.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+generatorName: scala-sttp4
+outputDir: samples/client/petstore/scala-sttp4-circe
```
Comment (wing328) [inline bin/configs/scala-sttp4-circe.yaml:2]: please add the new folder to the github workflow:

`.github/workflows/samples-scala-client.yaml`

File: samples/client/petstore/scala-sttp4/build.sbt
Comment (wing328) [inline samples/client/petstore/scala-sttp4/build.sbt]: do we need to update this in the template as well?

build failure: 

```
[info] set current project to openapi-client (in build file:/home/runner/work/openapi-generator/openapi-generator/samples/client/petstore/scala-sttp4/)
[error] java.lang.RuntimeException: Expected `scala-sttp4 / scalaVersion` to be 2.13.18 or later, but found 2.13.16.
[error] To support backwards-only binary compatibility (SIP-51), the Scala 2.13 compiler
[error] should not be older than scala-library on the dependency classpath.
[error] 
[error] Upgrade the `scalaVersion` to fix the build. If upgrading the Scala compiler version is
[error] not possible (for example due to a regression in the compiler or a missing dependency),
[error] this error can be demoted by setting `allowUnsafeScalaLibUpgrade := true`.
```

ref: https://github.com/OpenAPITools/openapi-generator/actions/runs/21925866201/job/63348973496?pr=22916

Comment (plaflamme) [reply samples/client/petstore/scala-sttp4/build.sbt]: Done in https://github.com/OpenAPITools/openapi-generator/pull/22916/commits/14227c6c0c6cf7546d445681dd0747cf1851c32d

File: bin/configs/scala-sttp4-circe.yaml (hunk 2 above)
Comment (plaflamme) [reply bin/configs/scala-sttp4-circe.yaml:2]: Done in https://github.com/OpenAPITools/openapi-generator/pull/22916/commits/545de9548dbd331797b91ea62ab2e590da56a803

**Comment by wing328** [comment]
please review the build failure when you've time.

let us know if you need any help

**Comment by plaflamme** [comment]
@wing328 I think this is good to go now?

**Comment by plaflamme** [comment]
@wing328 updated this on top of the latest master branch, anything else needed to get this in?

**Comment by wing328** [comment]
will try to run some tests tomorrow and merge if no questions/feedback from anyone.

File: modules/openapi-generator/src/test/resources/3_0/scala/sttp4-oneOf.yaml [hunk 3]
```diff
@@ -0,0 +1,107 @@
+openapi: 3.0.0
```
Comment (wing328) [inline modules/openapi-generator/src/test/resources/3_0/scala/sttp4-oneOf.yaml:1]: i did a test with this spec:

```
 java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g scala-sttp4 -i https://raw.githubusercontent.com/plaflamme/openapi-generator/ac1b84e6a05e1bbfbfed7903aed15cc20fdb7c9b/modules/openapi-generator/src/test/resources/3_0/scala/sttp4-oneOf.yaml -o /tmp/scala3 --additional-properties jsonLibrary=circe --skip-validate-spec
```

but got compilation errors on the output:

```
$ sbt compile
[info] welcome to sbt 1.10.11 (Eclipse Adoptium Java 11.0.30)
[info] loading settings for project scala3-build from plugins.sbt...
[info] loading project definition from C:\Users\User\AppData\Local\Temp\scala3\project
[info] loading settings for project scala3 from build.sbt...
[info] set current project to openapi-client (in build file:/C:/Users/User/AppData/Local/Temp/scala3/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] compiling 8 Scala sources to C:\Users\User\AppData\Local\Temp\scala3\target\scala-2.13\classes ...
[error] C:\Users\User\AppData\Local\Temp\scala3\src\main\scala\org\openapitools\client\core\JsonSupport.scala:19:66: not found: type DateSerializers
[error] object JsonSupport extends SttpCirceApi with AutoDerivation with DateSerializers with AdditionalTypeSerializers {
[error]                                                                  ^
[error] C:\Users\User\AppData\Local\Temp\scala3\src\main\scala\org\openapitools\client\model\Animal.scala:29:43: Could not find ConfiguredAsObjectEncoder for type org.openapitools.client.model.Animal.
[error] Some possible causes for this:
[error] - org.openapitools.client.model.Animal isn't a case class or sealed trait
[error] - some of org.openapitools.client.model.Animal's members don't have codecs of their own
[error] - missing implicit Configuration
[error]   implicit val encoder: Encoder[Animal] = deriveConfiguredEncoder
[error]                                           ^
[error] C:\Users\User\AppData\Local\Temp\scala3\src\main\scala\org\openapitools\client\model\Animal.scala:30:43: Could not find ConfiguredDecoder for type org.openapitools.client.model.Animal.
[error] Some possible causes for this:
[error] - org.openapitools.client.model.Animal isn't a case class or sealed trait
[error] - some of org.openapitools.client.model.Animal's members don't have codecs of their own
[error] - missing implicit Configuration
[error]   implicit val decoder: Decoder[Animal] = deriveConfiguredDecoder
[error]                                           ^
[error] C:\Users\User\AppData\Local\Temp\scala3\src\main\scala\org\openapitools\client\model\Pet.scala:22:40: could not find Lazy implicit value of type io.circe.generic.encoding.DerivedAsObjectEncoder[org.openapitools.client.model.Pet]
[error]   implicit val encoder: Encoder[Pet] = deriveEncoder
[error]                                        ^
[error] C:\Users\User\AppData\Local\Temp\scala3\src\main\scala\org\openapitools\client\model\Pet.scala:23:40: could not find Lazy implicit value of type io.circe.generic.decoding.DerivedDecoder[org.openapitools.client.model.Pet]
[error]   implicit val decoder: Decoder[Pet] = deriveDecoder
[error]                                        ^
[error] 5 errors found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 4 s, completed Mar 22, 2026, 10:09:05 PM
```

can you please take a look when you've time?

File: modules/openapi-generator/src/main/resources/scala-sttp4/model.mustache [hunk 4]
```diff
@@ -15,6 +20,296 @@ import {{import}}
...
+      case x: {{classname}} => Extraction.decompose(x)
+      {{/vendorExtensions.x-oneOfMembers}}
+      {{#vendorExtensions.x-wrappedOneOfMembers}}
+      case {{wrapperClassname}}(v) => Extraction.decompose(v)
+      {{/vendorExtensions.x-wrappedOneOfMembers}}
+    }
+  }
+{{/vendorExtensions.x-use-discr}}
+{{#vendorExtensions.x-use-discr}}
+  // oneOf with discriminator (with wrapper composition)
```
[the remark this answers is not in the document]
Comment (plaflamme) [reply modules/openapi-generator/src/main/resources/scala-sttp4/model.mustache:149]: Fixed in rebased commit

File: modules/openapi-generator/src/test/resources/3_0/scala/sttp4-oneOf.yaml (hunk 3 above)
Comment (plaflamme) [reply modules/openapi-generator/src/test/resources/3_0/scala/sttp4-oneOf.yaml:1]: @wing328 sorry for the delay, I've added https://github.com/OpenAPITools/openapi-generator/pull/22916/commits/5a3afd8083861627c56c09f61fb5b415d3184598 to fix this and merged the master branch to check everything is working, I think this is good to go again

**Comment by wing328** [comment]
did another tests and the output compiles without issue
```
[info] compiling 9 Scala sources to C:\Users\User\AppData\Local\Temp\scala4\target\scala-2.13\classes ...
[success] Total time: 6 s, completed Jun 5, 2026, 10:37:07 AM
```
